### PR TITLE
Generate Risor bindings in builtin modules

### LINKED ISSUE #158: Add code generator for writing Risor Go modules
Writing the Go modules for Risor is quite easy, but there's so much boilerplate. What if you wrote the functions as normal Go functions, and then generate the mapping functions to it?

Such as:

```go
//risor:export
func Repeat(str string, count int) string {
	return strings.Repeat(str, count)
}
```

And it generates a file with this:

```go
func gen_Repeat(ctx context.Context, args ...object.Object) object.Object {
	numArgs := len(args)
	if numArgs < 2 {
		return object.Errorf("type error: strings.repeat() takes 2 arguments (%d given)", len(args))
	}
	param1, err := object.AsString(args[0])
	if err != nil {
		return err
	}
	param2, err := object.AsInt(args[1])
	if err != nil {
		return err
	}
	return object.NewString(Repeat(param1, int(param2)))
}

func Module() *object.Module {
	return object.NewBuiltinsModule("strings", map[string]object.Object{
		"repeat": object.NewBuiltin("strings.repeat", gen_Repeat),
	})
}
```

This would not replace the docs generation and parsing proposed by https://github.com/risor-io/risor/issues/118#issuecomment-1859215640 but instead complement it. One does not exclude the other

### PR DESCRIPTION
**Pull request by applejag**
Writing Risor Go modules (aka builtin modules) is a lot of boilerplate.

This is a proposal to automate some parts of that, by writing our own Go code generator.

To start with, I've only implemented some core features, just so I can try it out. I've also replaced most functions in the `strings` module with these generated ones. To replace the remaining functions in the `strings` module too, I'd just have to add support for slice types, which isn't a biggy.

Some features that I did not implement just to keep the initial PR smaller:

- Map `error` returns and `(any, error)` into correct `object.NewError` return.
- Variadic function parameters `func(...any)`
- Bigger set of supported types in params and returns, such as `[]string`, `[]byte`, `map[string]any`, `SomeCustomType`
- Custom types and methods, such as the one from `exec.command`
- Parameters defaults

## Benefits

- Easier to write and maintain builtin modules

## Drawbacks

- Must run the `go generate` command after every edit. Could be paired with `entr` like in `make docs-dev`, and should be added to CI pipelines to ensure it's not forgotten.
- In its current form, the Go API of these modules changes. The public functions' signatures are changed from `func(context.Context, ...object.Object) object.Object` to whatever the function declares. Can be fixed by for example having the function with `//risor:export` being the private function.
- Adds another layer of abstraction, which can become tricky when you need fine-grained control. This solution aims to be opt-in on a per-function basis, so it should not be as intrusive.

---

Closes #158

### CODE REVIEW

**Comment by myzie** [comment]
This looks really promising!

> This solution aims to be opt-in on a per-function basis, so it should not be as intrusive.

I think this is a key point. This should be great as something that's _available_ but not required.

I'll review in more detail. Awesome stuff.

**Comment by luisdavim** [comment]
I haven't checked the code yet and I like the idea of having a tool to generate the boilerplate, like `cobra` and `kubebuilder` do but I'm not so sure about the drawbacks mentioned in the PR description, they seem pretty significant to me and sound like something that would get in the way of my workflow.

**Comment by applejag** [comment]
> I haven't checked the code yet and I like the idea of having a tool to generate the boilerplate, like `cobra` and `kubebuilder` do but I'm not so sure about the drawbacks mentioned in the PR description, they seem pretty significant to me and sound like something that would get in the way of my workflow.

I listed the drawback together with counterpoints to the drawbacks in the same bulletpoints.

Go's build system isn't super flexible, so forcing you to run `go generate` is pretty much the best we can do. It's also the approach taken by lots of other projects, such as gRPC, [templ](https://templ.guide/), and kubebuilder as you mentioned.

Regarding the layer of abstraction one, the counterpoint to that one is that if you want full control, then you just implement the function yourself without the generated mapping. The way it's done in this PR means that you can mix and match generated and non-generated functions freely in the same module.

**Comment by luisdavim** [comment]
I think a bit of duplication and some copy/pasting is better than adding complexity and forcing a specific workflow.
As for `kubebuilder` I was talking specifically about the boilerplate generator part, not the manifest and deep copy generation part and you are not required to run dose on every edit for sure.

I guess this case should be closer to what cobra does for adding new commands, it generates a starting point for you to edit and that's it, I think it would be quite simple to do that for a new module and no changes would be needed on the existing code for that.

**Comment by myzie** [comment]
It seems like the approach prototyped here doesn't _force_ a specific workflow, in that the module author can choose to opt-in using the `risor:generate` comment.

Do you agree @luisdavim?

File: cmd/risor-modgen/main.go [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+{{ range .BuildConstraints }}
+{{ . }}
+{{- end }}
+{{- end }}
+
+package {{ .Package }}
+
+import (
+	"context"
+	"math"
```
Comment (luisdavim) [inline cmd/risor-modgen/main.go:171]: Why import this if it's not being used?

**Review by luisdavim — COMMENTED** [review]
I think the generated code should be committed, and there should be a make target for triggering the code generation.

**Comment by luisdavim** [comment]
> It seems like the approach prototyped here doesn't _force_ a specific workflow, in that the module author can choose to opt-in using the `risor:generate` comment.
> 
> Do you agree @luisdavim? 

Yes, looks like it.
I still think that something that would generate a starting point to be edited would be simpler and more useful.

This approach looks very cool but I think it will end up serving more limited use cases.

File: cmd/risor-modgen/main.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+	"github.com/risor-io/risor/object"
+)
+
+// Stop compile error from unused imports
+var _ context.Context
```
Comment (luisdavim) [inline cmd/risor-modgen/main.go:177]: I don't think we should do this.

Comment (applejag) [reply cmd/risor-modgen/main.go:177]: I didnt want to go into tracking imports yet in this PR, so I added this hack instead as context.Context is only sometimes used.

File: cmd/risor-modgen/main.go (hunk 1 above)
Comment (applejag) [reply cmd/risor-modgen/main.go:171]: It's used if you have for example an int parameter. The object.AsInt returns a int64, so to convert to int we first need to check if the value is not bigger than math.MaxInt

File: modules/strings/strings.go [hunk 3]
```diff
@@ -15,63 +18,29 @@ func asString(obj object.Object) (*object.String, *object.Error) {
...
-		return err
-	}
-	s, err := asString(args[0])
-	if err != nil {
-		return err
-	}
-	return s.Contains(args[1])
+//risor:export
+func Contains(s, substr string) bool {
+	return strings.Contains(s, substr)
```
Comment (myzie) [inline modules/strings/strings.go:23]: Would it be possible to use this form?

```
var Contains = strings.Contains
```

Presumably the Go AST is different, but maybe this case could be handled?

Comment (applejag) [reply modules/strings/strings.go]: Yeah the AST doesn't contain type information.

The approach would be to use [`go/types`](https://pkg.go.dev/go/types) to parse the type info, and potentially combine that with [`go/importer`](https://pkg.go.dev/go/importer). And then from that resolve the function definition of `strings.Contains`.

Certainly doable, but I'd say better to leave that for a follow-up PR.

File: cmd/risor-modgen/main.go (hunk 2 above)
Comment (applejag) [reply cmd/risor-modgen/main.go]: I added import tracking. So now the `Stop compile error` section is gone.

File: cmd/risor-modgen/main.go (hunk 1 above)
Comment (applejag) [reply cmd/risor-modgen/main.go]: I added import tracking. So now the `Stop compile error` section is gone.

**Comment by applejag** [comment]
> I think the generated code should be committed, and there should be a make target for triggering the code generation.

The generated code is committed, and I've now added two targets:

- `make generate` to just run `go generate`
- `make modgen` to run `./cmd/risor-modgen` via `entr`

**Pull request merged**